From 7f1f1425b0f105bd51d993df944bc34f80ca3989 Mon Sep 17 00:00:00 2001
From: Alexey Makhalov <amakhalov@vmware.com>
Date: Tue, 4 Aug 2020 01:21:43 +0000
Subject: [PATCH] tcp: inherit TSQ limit from root namespace
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

It will allow conttainer host to use non standard TSQ limit
in all container namespaces by setting system wide property
once.

We do not want to increase default in-kernel value from
256KB to 1MB as it was done upstream in 5.0 in
c73e5807e4f6fc6d373a5db55b45f639f8bb6328
(“tcp: tsq: no longer use limit_output_bytes for paced flows”)
as it might depend on other changes by Eric Dumazet around
this time, such as
ab408b6dc7449c0f791e9e5f8de72fa7428584f2
(“tcp: switch tcp and sch_fq to new earliest departure time model”)

Increasing default value to 1MB shows better single socket
throughput (by iperf) over NSX networking, but at the same
time it can make worse bufferbloat situation.

Let customer to pick best value for him.
---
 net/ipv4/tcp_ipv4.c | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/net/ipv4/tcp_ipv4.c b/net/ipv4/tcp_ipv4.c
index 7536f4c0bbf4..3d4d1b6ac56a 100644
--- a/net/ipv4/tcp_ipv4.c
+++ b/net/ipv4/tcp_ipv4.c
@@ -2580,8 +2580,6 @@ static int __net_init tcp_sk_init(struct net *net)
 	 * which are too large can cause TCP streams to be bursty.
 	 */
 	net->ipv4.sysctl_tcp_tso_win_divisor = 3;
-	/* Default TSQ limit of four TSO segments */
-	net->ipv4.sysctl_tcp_limit_output_bytes = 262144;
 	/* rfc5961 challenge ack rate limiting */
 	net->ipv4.sysctl_tcp_challenge_ack_limit = 1000;
 	net->ipv4.sysctl_tcp_min_tso_segs = 2;
@@ -2597,6 +2595,12 @@ static int __net_init tcp_sk_init(struct net *net)
 		memcpy(net->ipv4.sysctl_tcp_wmem,
 		       init_net.ipv4.sysctl_tcp_wmem,
 		       sizeof(init_net.ipv4.sysctl_tcp_wmem));
+		/* Inherit system wide TSQ limit */
+		net->ipv4.sysctl_tcp_limit_output_bytes =
+			init_net.ipv4.sysctl_tcp_limit_output_bytes;
+	} else {
+		/* Default TSQ limit of four TSO segments */
+		net->ipv4.sysctl_tcp_limit_output_bytes = 262144;
 	}
 	net->ipv4.sysctl_tcp_comp_sack_delay_ns = NSEC_PER_MSEC;
 	net->ipv4.sysctl_tcp_comp_sack_nr = 44;
-- 
2.14.2

